common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,305 @@
+    return FALSE;
+
+  return TRUE;
+}
+
+#define EVENT_LOG_VERSION "1"
+
+static JsonNode *
+build_base_event (const gchar *event)
+{
+  g_autoptr(JsonBuilder) builder = json_builder_new ();
+
+  json_builder_begin_object (builder);
+
+  json_builder_set_member_name (builder, "event");
+  json_builder_add_string_value (builder, event);
+
+  json_builder_set_member_name (builder, "version");
+  json_builder_add_string_value (builder, EVENT_LOG_VERSION);
+  json_builder_end_object (builder);
+
+  return json_builder_get_root (builder);
+}
+
+static JsonNode *
+deploy_install_event (const gchar *ref,
+                      const gchar *origin)
+{
+  JsonNode *base = build_base_event ("deploy-install");
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+  json_object_set_string_member (object, "origin", origin);
+
+  return base;
+}
+
+gboolean
+flatpak_transaction_log_commit_deploy_install_event (FlatpakTransactionLog *log,
+		                                     const gchar           *ref,
+		                                     const gchar           *origin,
+		                                     GCancellable          *cancellable,
+		                                     GError                **error)
+{
+  g_autoptr(JsonNode) node = load_existing_log_if_exists (log->path, cancellable, error);
+  JsonArray *array = json_node_get_array (node);
+  g_autoptr(JsonNode) event = deploy_install_event (ref, origin);
+
+  json_array_add_object_element (array, json_node_dup_object (event));
+
+  return write_node_to_log_file (log->path, node, cancellable, error);
+}
+
+static JsonNode *
+uninstall_event (const gchar *ref)
+{
+  JsonNode *base = build_base_event ("uninstall");
+  JsonObject *object = json_node_get_object (base);
+
+  json_object_set_string_member (object, "ref", ref);
+
+  return base;
+}
+
+gboolean
+flatpak_transaction_log_commit_uninstall_event (FlatpakTransactionLog *log,
+                                                const gchar           *ref,
+                                                GCancellable          *cancellable,
+                                                GError                **error)
+{
+  g_autoptr(JsonNode) node = load_existing_log_if_exists (log->path, cancellable, error);
+  JsonArray *array = json_node_get_array (node);
+  g_autoptr(JsonNode) event = uninstall_event (ref);
+
+  json_array_add_object_element (array, json_node_dup_object (event));
+
+  return write_node_to_log_file (log->path, node, cancellable, error);
+}
+
+static void
+flatpak_transaction_log_class_init (FlatpakTransactionLogClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = flatpak_transaction_log_finalize;
+  object_class->get_property = flatpak_transaction_log_get_property;
+  object_class->set_property = flatpak_transaction_log_set_property;
+
+  g_object_class_install_property (object_class,
```

Comment:
Add a gtk-doc comment for it (and for the rest of the API in the new class). I realise it’s not public API, but documentation is useful for whoever’s editing things inside flatpak.git too.